### ISSUE TITLE
use softmax when inference, cross_entropy when training

### ISSUE THREAD

**Issue by undertherain** [undertherain/benchmarker#31]
When training DNNs, softmax is typically used in log form and fused with  negative log likelihood loss - for performance and numerical stability. 
I.e. the output of the model is logits

When doing inference, at least for  classification models, we typically need to apply softmax on logits to get normalized probabilities

Some of the models we use come with softmax pre-backed in them, or sometimes log-sofmax. This makes unified implementation harder and currently we have a bit of a mess with this.

**Comment by vatai** [comment]
[This](https://pytorch.org/docs/stable/nn.html#crossentropyloss) is what we are talking about, right?

So what we'd expect is to have something with this logic (in principle):
```python
if params["mode"] == "inference":
  return softmax_model # with logits -> softmax
else: # if params["mode"] == "training"
  return cross_entropy_loss_model # with logits -> cross_entropy_loss
  # which is the same as logits -> {{{softmax -> log -> NLL_loss}}}
```
right?

**Comment by vatai** [comment]
copied from #44 

# Proposed solution
The `self.net` member of `INeuralNet` stores the model returned by `get_kernel()` of the problem (e.g. `get_kernel()` defined in `pytorch.py`). Instead of putting just the model in `self.net`, we could add the model and the loss as well... i.e. so all the pytorch `get_kernel` functions should `return (model, loss)` and `do_python.py` should use `self.net[0]` instead of `self.net` and `self.net[1]` instead of `CrossEntropyLoss()`

**Comment by undertherain** [comment]
we should not return model and loss - because for training the loss often fused with last layer activation

**Comment by undertherain** [comment]
better return two versions of a model

**Comment by vatai** [comment]
Wait, I'm not so good with pytorch, but don't you need a "loss object", and use it like in the code below (our current code)? 
```python
            criterion = nn.CrossEntropyLoss()
            loss = criterion(output, target)
            loss.backward()
            optimizer.step()
            log_interval = 10
            if batch_idx % log_interval == 0:
                print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
                    epoch, batch_idx * len(data), len(self.x_train),
                    100. * batch_idx / len(self.x_train), loss.item()))
```

So, in the current code we have the combined/fused situation, no?

PyTorch [docs](https://pytorch.org/docs/master/generated/torch.nn.CrossEntropyLoss.html) say:

> This criterion combines nn.LogSoftmax() and nn.NLLLoss() in one single class.

And of course this is pytorch exclusive. Other frameworks will need to handle this differently (e.g. TF has the loss saved in the model). No?

And what you say: 
> we should not return model and loss - because for training the loss often fused with last layer activation

Isn't this be exactly the reason why we need to return the loss too... i.e. as I understand the loss  (in case of training) is (logically) part of the model... for inference we can simply ignore the loss (return `None` and not use it). - and again... this is PyTorch exclusive!

**Comment by undertherain** [comment]
you conflate everything in one message:

> but don't you need a "loss object"
that's an implementation detail which is not important. can do loss as layer, as object, as function call - does not matter

>Isn't this be exactly the reason why we need to return the loss too.
it is the reason why we _should_not_ return loss in addition to the model

>for inference we can simply ignore the loss
we can't because we need to add softmax instead, in case of classification

**Comment by vatai** [comment]
I know (I assumed) the loss was a technical detail, but I guess I lack pytorch experience. If you could show me other ways how you can implement the loss (send link or stg) I'd appreciate it! onegaitashimasu!

**Comment by vatai** [comment]
@undertherain this can be closed right? Please close if yes!
